### ISSUE TITLE
Gateway listener port conflict when creating multiple k8s services

### ISSUE THREAD

**Issue by gitanuj** [istio/istio#46820]
I'm running into an issue for which there's a [hotfix](https://github.com/istio/istio/pull/33021) available however it specifically says that it is meant only for KNative and should not be used. So I'm looking for guidance on how to create Gateway in such scenario.

I'm creating following Istio resources -
- **Istio Gateway** with 2 server ports **A** and **B**
- **K8s Service** with **port A** and **targetPort A**
- **K8s Service** with **port A** and **targetPort B**

Now depending on order of creation of K8s Service envoy ends up with different listener ports because of how Istio resolves ports. A [workaround](https://github.com/istio/istio/pull/33021) was merged to fix this specifically for KNative.

I need to create 2 different K8s Service because I want to provision 2 different cloud LBs (with different networking setup) but I want to route traffic to the same gateway pods to share the resources. I also want to have 2 different listeners in Envoy which have slightly different filter chains because of different networking. Is there a way I can ensure I don't run into the gateway port conflict while being able to create multiple K8s services pointing to the same Gateway?

**Comment by hzxuzhonghu** [comment]
@howardjohn

**Comment by hzxuzhonghu** [comment]
Can you explain how these two ports [A, B] conflicts

**Comment by howardjohn** [comment]
It conflicts because Istio has some logic around the service and target ports. The Gateway port is supposed to actually apply to the Service port, so when its applied to the target port things get mixed up.

**Comment by howardjohn** [comment]
It works fine with Gateway API fwiw

**Comment by hzxuzhonghu** [comment]
We have changed to use gateway service target port in higher priority, so if A and B are different why should they conflict?

But service port P can target to A/B, I am not sure how your load balancer can handle that

**Comment by hzxuzhonghu** [comment]
Better elaborate a bit more about the complete ingress traffic topology

**Comment by gitanuj** [comment]
I want the below networking setup. I need 2 LoadBalancer Services with same customer facing port `3333` but internally in Envoy I need the traffic to be received by different ports `3333` and `3334`.

![image](https://github.com/istio/istio/assets/693092/493d183d-c7b3-42e1-a7db-1e1637d14f31)

To accomplish this I installed the Istio Ingress Gateway helm chart (which created a LoadBalancer Service) and I created the Gateway and an additional LoadBalancer Service. So I ended up with below resources -

```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: my-gateway
  namespace: my-namespace
spec:
  selector:
    app: istio-gateway
    istio: gateway
  servers:
    - hosts:
        - my-namespace/*
      port:
        name: my-port
        number: 3333
        protocol: TLS
      tls:
        mode: PASSTHROUGH
    - hosts:
        - my-namespace/*
      port:
        name: my-port-private
        number: 3334
        protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
    service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
    service.beta.kubernetes.io/aws-load-balancer-type: external
  labels:
    app: istio-gateway
    istio: gateway
  name: istio-gateway
  namespace: my-namespace
spec:
  ports:
    - name: my-port
      port: 3333
      protocol: TCP
      targetPort: 3333
  selector:
    app: istio-gateway
    istio: gateway
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
    service.beta.kubernetes.io/aws-load-balancer-scheme: internal
    service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: proxy_protocol_v2.enabled=true
    service.beta.kubernetes.io/aws-load-balancer-type: external
  labels:
    app: istio-gateway-private
    istio: gateway-private
  name: istio-gateway-private
  namespace: my-namespace
spec:
  ports:
    - name: my-port-private
      port: 3333
      protocol: TCP
      targetPort: 3334
  selector:
    app: istio-gateway
    istio: gateway
  type: LoadBalancer
```

However this doesn't work correctly if `istio-gateway-private` Service is created first and the other Service is created second. I end up with just port `3334` as a dynamic_listener on envoy (however I expected both ports `3333` and `3334` to be listening).

**Comment by hzxuzhonghu** [comment]
Thanks, it's clear to me

**Comment by gitanuj** [comment]
As a workaround I ended up installing another Istio Ingress Gateway (with a new set of Deployment, Service and Gateway altogether). Would be great if I could share the same envoy Deployment for the above scenario though.

**Comment by howardjohn** [comment]
> Istio Gateway with 2 server ports A and B
> K8s Service with port P and targetPort A
> K8s Service with port P and targetPort B

If you actually did this it would work fine

The problem mis you have

Istio Gateway with 2 server ports A and B
K8s Service with port **A** and targetPort **A**
K8s Service with port **A** and targetPort B

So **A** is ambiguous.

Just make the target ports different (and refer to target port in Gateway) and it should work fine

**Comment by gitanuj** [comment]
Unfortunately we are already using Istio Ingress Gateway in production with the following config
- Istio Gateway with one **server port A**
- K8s Service with **port A** and **targetPort A**

And I want to add another K8s Service with **port A** and **targetPort B** while maintaining backwards compatibility. (Otherwise I'll need to have some migration). So for now just using a new Istio Ingress Gateway deployment seems to work.

**Comment by gitanuj** [comment]
I just tested this quickly with an already configured Gateway (with multiple VirtualServices for routes) -

#### initial state
- Istio Gateway with one server **port A**
- K8s Service with **port A** and **targetPort A**

#### new state
- Istio Gateway with one server **port A**
- K8s Service with **port A** and **targetPort X**

The only thing I changed is Service targetPort and I expected that envoy will be configured to listen on port X instead. However that didn't happen. Envoy still listened on old **port A** and health checks for my LB started failing because it started using **port X**. 

However a new deployment works fine and envoy listens on **port X** as expected. May be Istio doesn't react to changes in Service targetPort?

**Comment by gitanuj** [comment]
I tested this by updating the Gateway server port and Service targetPort to a new value but it looks like I will also need to update a bunch of VirtualServices that we have because we also use port number for matching. So that means I'll need to migrate all my VirtualServices, Gateway and Service to use the new port.

That could work but it comes with risks of migration and having a new Istio Gateway deployment is acceptable workaround for us.

**Comment by hzxuzhonghu** [comment]
Confused， IMO gateway service port have nothing to do with VS

**Comment by gitanuj** [comment]
I'm using `tls.match.port` which I'm guessing is also relying on gateway server port matching and thus making it difficult to change the Gateway server port without migrating all the Virtual Services too.

**Comment by bvandewalle** [comment]
I know this ticket is closed but we had a similar issue today.
We updated our service Type Loadbalancer as:
User-->443 -->NLB-->8443-->Envoy

But this started to work before we updated our Envoy Listening port to 8443 on the Gateway CRDs.

after a lot of investigation and reading the code, we realized that Istiod was acting as a controller and trying to play smart by resolving the Gateway CRD Port to the public facing port, not the TargetPort.

This bit us on some clusters because we had multiple services some with targetPort as 443 and some as 8443.

I'm going to open a standalone ticket for this.
